hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/onset/OnsetChecker.java

```diff
@@ -0,0 +1,230 @@
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import java.time.Instant;
+
+/** This class preprocess transactions by parsing them and checking for syntax errors. */
+public class OnsetChecker {
+
+    private final int maxSignedTxnSize;
+    private final RecordCache recordCache;
+    private final GlobalDynamicProperties dynamicProperties;
+    private final HapiOpCounters counters;
+
+    /**
+     * Constructor of an {@code OnsetChecker}
+     *
+     * @param maxSignedTxnSize the maximum transaction size
+     * @param recordCache the {@link RecordCache}
+     * @param dynamicProperties the {@link GlobalDynamicProperties}
+     * @param counters metrics related to workflows
+     * @throws NullPointerException if one of the arguments is {@code null}
+     */
+    public OnsetChecker(
+            final int maxSignedTxnSize,
+            @NonNull final RecordCache recordCache,
+            @NonNull final GlobalDynamicProperties dynamicProperties,
+            @NonNull final HapiOpCounters counters) {
+        if (maxSignedTxnSize <= 0) {
+            throw new IllegalArgumentException("maxSignedTxnSize must be > 0");
+        }
+        this.maxSignedTxnSize = maxSignedTxnSize;
+        this.recordCache = requireNonNull(recordCache);
+        this.dynamicProperties = requireNonNull(dynamicProperties);
+        this.counters = requireNonNull(counters);
+    }
+
+    /**
+     * Validates a {@link Transaction}
+     *
+     * @param tx the {@code Transaction}
+     * @throws PreCheckException if validation fails
+     * @throws NullPointerException if {@code tx} is {@code null}
+     */
+    @SuppressWarnings("deprecation")
+    public void checkTransaction(@NonNull final Transaction tx) throws PreCheckException {
+        requireNonNull(tx);
+
+        final var hasSignedTxnBytes = !tx.getSignedTransactionBytes().isEmpty();
+        final var hasDeprecatedSigMap = tx.hasSigMap();
+        final var hasDeprecatedBodyBytes = !tx.getBodyBytes().isEmpty();
+        final var hasDeprecatedBody = tx.hasBody();
+        final var hasDeprecatedSigs = tx.hasSigs();
+
+        if (hasDeprecatedBody
+                || hasDeprecatedSigs
+                || hasDeprecatedSigMap
+                || hasDeprecatedBodyBytes) {
+            counters.countDeprecatedTxnReceived();
+        }
+
+        if (hasSignedTxnBytes) {
+            if (hasDeprecatedBodyBytes || hasDeprecatedSigMap) {
+                throw new PreCheckException(INVALID_TRANSACTION);
+            }
+        } else if (!hasDeprecatedBodyBytes) {
+            throw new PreCheckException(INVALID_TRANSACTION_BODY);
+        }
+
+        if (tx.getSerializedSize() > maxSignedTxnSize) {
+            throw new PreCheckException(TRANSACTION_OVERSIZE);
+        }
+
+        if (MiscUtils.hasUnknownFields(tx)) {
+            throw new PreCheckException(TRANSACTION_HAS_UNKNOWN_FIELDS);
+        }
```

Comment:
This one is interesting, we need to make sure pbj has support for this (currently it ignores unknown fields). We should also document why this is important, because I found it a surprising requirement. If we don't have this check, it would mean transactions from a newer client to an older network could submit transactions, which is probably surprising, so this check is probably a very good thing. Also people could send arbitrary protobuf fields and we'd put them in state and in the record stream, which might also be surprising.